### PR TITLE
Sequence Number Panic

### PR DESCRIPTION
**Pull request by l-monninger**
# Summary
- **RFCs**: $\emptyset$
- **Categories**: `protocol-units`

Updates sequence number handling in mempool to not panic executor thread. 

# Changelog

1. Introduces `thiserror` enum for `transaction_pipe`.
2. Catches `MempoolStatus::NotAccepted` in `dof-executor`. 

# Testing

1. e2e tests. 

# Outstanding issues
1. Error types should be introduced elsewhere.

### CODE REVIEW

File: protocol-units/execution/dof/src/v1.rs [hunk 1]
```diff
@@ -50,7 +51,17 @@ impl DynOptFinExecutor for Executor {
...
 		loop {
 			// readers should be able to run concurrently
-			self.executor.tick_transaction_pipe(self.transaction_channel.clone()).await?;
+			match self.executor.tick_transaction_pipe(self.transaction_channel.clone()).await {
+				Ok(_) => {}
+				Err(e) => match e {
+					TransactionPipeError::TransactionNotAccepted(e) => {
+						// allow the transaction not to be accepted by the mempool
+						// because the client may have sent a bad sequence number
+						tracing::error!("Transaction not accepted: {:?}", e);
```
Comment (mzabaluev) [inline protocol-units/execution/dof/src/v1.rs:60]: The error severity level is meant for fatal things. If this is to be handled in the normal course of operation, but logged as per default log settings, this should be `info!` or maybe `warn!`.

File: protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs [hunk 2]
```diff
@@ -37,10 +54,11 @@ impl Executor {
 
 										match status.code {
 											MempoolStatusCode::Accepted => {
-
+												tracing::debug!("Transaction accepted: {:?}", transaction);
 											},
 											_ => {
-												anyhow::bail!("Transaction not accepted: {:?}", status);
+												tracing::debug!("Transaction not accepted: {:?}", status);
```
Comment (mzabaluev) [inline protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs:60]: This is likely redundant with the log message that I commented on above.

File: protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs [hunk 3]
```diff
@@ -158,6 +176,56 @@ mod tests {
...
+			.send(MempoolClientRequest::SubmitTransaction(user_transaction.clone(), req_sender))
+			.await?;
+
+		// tick the transaction pipe
+		executor.tick_transaction_pipe(tx).await?;
+		/*match executor.tick_transaction_pipe(tx).await {
+			Err(TransactionPipeError::TransactionNotAccepted(_)) => {}
+			Err(e) => return Err(anyhow::anyhow!("Unexpected error: {:?}", e)),
+			Ok(_) => return Err(anyhow::anyhow!("Expected error")),
+		}*/
```
Comment (mzabaluev) [inline protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs:219]: Remove if no longer needed.

**Review by mzabaluev — APPROVED** [review]
Looks good, I've only got small remarks as usual.

File: protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs [hunk 4]
```diff
@@ -158,6 +176,56 @@ mod tests {
 		Ok(())
 	}
 
+	#[tokio::test]
+	async fn test_pipe_mempool_with_malformed_transaction() -> Result<(), anyhow::Error> {
```
Comment (l-monninger) [inline protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs:180]: @mzabaluev This should error. I don't know why it doesn't. It's two of the same transaction being submitted to the mempool.

File: protocol-units/execution/dof/src/v1.rs (hunk 1 above)
Comment (l-monninger) [reply protocol-units/execution/dof/src/v1.rs:60]: Sorry, yes, I'll change to `warn!`.

File: protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs (hunk 3 above)
Comment (l-monninger) [reply protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs:219]: Ideally, this gets fixed and goes in.

File: protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs (hunk 4 above)
Comment (mzabaluev) [reply protocol-units/execution/opt-executor/src/executor/transaction_pipe.rs:195]: Use `#[ignore]` as a maintenance-friendly way to mark a test that does not do what's expected, so it sticks out to anyone who would look at the test summaries later.

**Pull request merged**